File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9P.scala
package TextbookTrack.Chapter09

object BA9P {
  import scala.annotation.tailrec

  sealed trait NodeColor {
    override def toString: String = this match {
      case Blue => "blue"
      case Red => "red"
      case Purple => "purple"
    }
  }

  case object Blue extends NodeColor
  case object Red extends NodeColor
  case object Purple extends NodeColor

  object NodeColor {
    def apply(s: String): NodeColor = s match {
      case "blue" => Blue
      case "red" => Red
      case "purple" => Purple
      case _ => throw new Exception("Unknown node color.")
    }
  }

  private def readEdges(lines: Iterator[String]): Map[Int, List[Int]] =
    lines
      .takeWhile(_ != "-")
      .map(_.split(" -> ").toList)
      .collect{
        case List(node, neighbours) if neighbours != "{}" => (node.toInt, neighbours.split(",").map(_.toInt).toList)
      }
      .toMap

  private def readLeafColors(lines: Iterator[String]): Map[Int, NodeColor] =
    lines
      .map(_.split(": ").toList)
      .collect{ case List(node, color) => (node.toInt, NodeColor(color)) }
      .toMap

  private def collectRipeNodes(adjacencyList: Map[Int, List[Int]], nodeColors: Map[Int, NodeColor]): List[Int] = {
    (for {
      (node, children) <- adjacencyList
      if !nodeColors.contains(node) && children.forall(nodeColors.contains)
    } yield node).toList
  }

  private def assignColorToNode(node: Int,
                                adjacencyList: Map[Int, List[Int]],
                                nodeColors: Map[Int, NodeColor]): NodeColor = {
    val children: List[Int] = adjacencyList(node)
    val childColors: Set[NodeColor] = children.map(nodeColors).toSet
    if (childColors.size == 1) childColors.head
    else Purple
  }

  def performNodeColoring(adjacencyList: Map[Int, List[Int]], leafColors: Map[Int, NodeColor]): Map[Int, NodeColor] = {
    @tailrec
    def loop(nodeColors: Map[Int, NodeColor]): Map[Int, NodeColor] = collectRipeNodes(adjacencyList, nodeColors) match {
      case Nil => nodeColors
      case ripeNodes =>
        val coloredRipeNodes: List[(Int, NodeColor)] =
          ripeNodes.map{ node => (node, assignColorToNode(node, adjacencyList, nodeColors)) }
        loop(nodeColors ++ coloredRipeNodes)
    }

    if (adjacencyList.isEmpty) leafColors else loop(leafColors)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val adjacencyList: Map[Int, List[Int]] = readEdges(reader)
    val nodeColors: Map[Int, NodeColor] = readLeafColors(reader)
    val result: Map[Int, NodeColor] = performNodeColoring(adjacencyList, nodeColors)
    result.foreach { case (node, color) => println(s"$node: $color") }
  }
}

File: Python/textbook_track/chapter09/ba9p.py
# Implement TreeColoring
import sys

from enum import Enum


class NodeColor(Enum):
    BLUE = 'blue'
    RED = 'red'
    PURPLE = 'purple'


def read_edges(reader):
    adjacency_list = {}
    while (line := next(reader).rstrip()) != '-':
        node, neighbours = line.split(' -> ')
        if neighbours != '{}':
            adjacency_list[int(node)] = tuple(map(int, neighbours.split(',')))
    return adjacency_list


def read_leaf_colors(reader):
    node_colors = {}
    for line in reader:
        node, color = line.split(': ')
        node_colors[int(node)] = NodeColor(color.rstrip())
    return node_colors


def collect_ripe_nodes(adjacency_list, node_colors):
    ripe_nodes = []
    for node, children in adjacency_list.items():
        if node not in node_colors and all(child in node_colors for child in children):
            ripe_nodes.append(node)
    return ripe_nodes


def assign_color_to_node(node, adjacency_list, node_colors):
    children = adjacency_list[node]
    child_colors = {node_colors[child] for child in children}
    if len(child_colors) == 1:
        return next(iter(child_colors))

    return NodeColor.PURPLE


def perform_node_coloring(adjacency_list, node_colors):
    if not adjacency_list:
        return node_colors

    while ripe_nodes := collect_ripe_nodes(adjacency_list, node_colors):
        for node in ripe_nodes:
            node_colors[node] = assign_color_to_node(node, adjacency_list, node_colors)
    return node_colors


def main():
    reader = sys.stdin
    adjacency_list = read_edges(reader)
    node_colors = read_leaf_colors(reader)
    result = perform_node_coloring(adjacency_list, node_colors)
    for node, color in result.items():
        print(f'{node}: {color.value}')


if __name__ == '__main__':
    main()
